File: 21-1.rb
#!/usr/bin/env ruby

$monkeys = {}

class Monkey
  def initialize(name, operation)
    number_format = /^(?<num>[[:digit:]]+)$/
    calc_format = /^(?<m1>[[:alpha:]]+) (?<op>[-+*\/]) (?<m2>[[:alpha:]]+)$/

    @operation = operation
    @name = name
    @value = nil

    if m = number_format.match(@operation)
      @value = m['num'].to_i
    else
      m = calc_format.match(@operation)
      @m1 = m['m1']
      @op = m['op']
      @m2 = m['m2']
    end
  end

  def value
    return @value if @value

    @value = $monkeys[@m1].value.send(@op, $monkeys[@m2].value)
  end

  def to_s
    "<#{self.class}: #{@name} = #{@operation} (#{@value})>"
  end

  alias inspect to_s
end


input = File.read('21.input').lines

monkey_format = /^(?<name>[[:alpha:]]+): (?<op>.*)$/

input.each do |line|
  m = monkey_format.match line
  $monkeys[m['name']] = Monkey.new(m['name'], m['op'])
end

print $monkeys['root'].value, "\n"

File: 21-1.py
#!/usr/bin/env python

import re

monkeys = {}

class Monkey:
    def __init__(self, name, operation):
        number_format = re.compile("^(?P<num>[0-9]+)$")
        calc_format = re.compile("^(?P<m1>[a-z]+) (?P<op>[-+*/]) (?P<m2>[a-z]+)$")

        self.operation = operation
        self.name = name
        self.val = None
        if (m := number_format.match(self.operation)):
            self.val = int(m["num"])
        else:
            m = calc_format.match(self.operation)
            self.m1 = m["m1"]
            self.op = m["op"]
            self.m2 = m["m2"]

    def value(self):
        if self.val is not None:
            return self.val

        if self.op == "+":
            self.val = monkeys[self.m1].value() + monkeys[self.m2].value()
        elif self.op == "-":
            self.val = monkeys[self.m1].value() - monkeys[self.m2].value()
        elif self.op == "*":
            self.val = monkeys[self.m1].value() * monkeys[self.m2].value()
        elif self.op == "/":
            self.val = monkeys[self.m1].value() // monkeys[self.m2].value()

        return self.val

    def __str__(self):
        return f"<{self.__class__.__name__}: {self.name} = {self.operation} ({self.val})>"


monkey_format = re.compile("^(?P<name>[a-z]+): (?P<op>.*)$")

lines = [row.strip() for row in open("21.input").readlines()]

for line in lines:
    m = monkey_format.match(line)
    monkeys[m["name"]] = Monkey(m["name"], m["op"])

print(monkeys["root"].value())
